// openfsm-common/Common.js
/* Общая библиотека */

/* Paths */
const path = require("path");
const COMMON_PATH_TO_SITE = path.join(path.resolve(), 'project'); // путь до корня сайта
const COMMON_PATH_TO_PAGES = path.join(COMMON_PATH_TO_SITE, 'src', 'pages'); // путь до страниц-шаблонов
const COMMON_PATH_TO_PARTIALS = path.join(COMMON_PATH_TO_PAGES, 'partials'); // путь до блоков

/* Pages */
const COMMON_REGISTRATION_PAGE = path.join(COMMON_PATH_TO_PAGES, 'registration.hbs'); // Путь до страницы регистрации
const COMMON_REGISTRATION_CONFIRM_PAGE = path.join(COMMON_PATH_TO_PAGES, 'registration-confirm.hbs'); // Путь до страницы подтверждения регистрации
const COMMON_REGISTRATION_SUCCESS_PAGE = path.join(COMMON_PATH_TO_PAGES, 'registration-success.hbs'); // Путь до страницы успешной регистрации
const COMMON_REGISTRATION_DECLINE_PAGE = path.join(COMMON_PATH_TO_PAGES, 'registration-decline.hbs'); // Путь до страницы отклоненной регистрации
const COMMON_REGISTRATION_FAILURE_PAGE = path.join(COMMON_PATH_TO_PAGES, 'registration-failure.hbs'); // Путь до страницы ошибки регистрации
const COMMON_LOGON_PAGE = path.join(COMMON_PATH_TO_PAGES, 'logon.hbs'); // Путь до страницы входа
const COMMON_LOGOUT_PAGE = path.join(COMMON_PATH_TO_PAGES, 'logout.hbs'); // Путь до страницы выхода
const COMMON_OUT_SERVICE_PAGE = path.join(COMMON_PATH_TO_PAGES, 'out-service.hbs'); // Путь до страницы временной недоступности
const COMMON_SESSION_CLOSE_PAGE = path.join(COMMON_PATH_TO_PAGES, 'session-close.hbs'); // Путь до страницы закрытия сессии
const COMMON_APP_PAGE = path.join(COMMON_PATH_TO_PAGES, 'app.hbs'); // Путь до страницы приложения
const COMMON_PRODUCTS_PAGE = path.join(COMMON_PATH_TO_PAGES, 'app.hbs'); // Путь до витрины
const COMMON_PROFILE_PAGE = path.join(COMMON_PATH_TO_PAGES, 'app.hbs'); // Путь до профиля
const COMMON_BASKET_PAGE = path.join(COMMON_PATH_TO_PAGES, 'app.hbs'); // Путь до корзины
const COMMON_ORDERS_PAGE = path.join(COMMON_PATH_TO_PAGES, 'app.hbs'); // Путь до заказов
const COMMON_GET_ORDER_PAGE = path.join(COMMON_PATH_TO_PAGES, 'app.hbs'); // Путь до страницы получения заказа
const COMMON_GET_ORDER_SUCCESS_PAGE = path.join(COMMON_PATH_TO_PAGES, 'app.hbs'); // Путь до страницы успешного получения заказа
const COMMON_GET_ORDER_ERROR_PAGE = path.join(COMMON_PATH_TO_PAGES, 'app.hbs'); // Путь до страницы ошибки получения заказа
const COMMON_404_PAGE = path.join(COMMON_PATH_TO_PAGES, 'page-404.hbs'); // Путь до страницы 404

/* HTTP-codes */                
const HTTP_CODES = {
    OK: {
        code: 200,
        description: "Успешный запрос."
    },
    CREATED: {
        code: 201,
        description: "Ресурс успешно создан."
    },
    ACCEPTED: {
        code: 202,
        description: "Запрос принят, но ещё не обработан."
    },
    NO_CONTENT: {
        code: 204,
        description: "Успешный запрос, но тело ответа пусто."
    },
    BAD_REQUEST: {
        code: 400,
        description: "Некорректный запрос."
    },
    UNAUTHORIZED: {
        code: 401,
        description: "Необходима аутентификация."
    },
    FORBIDDEN: {
        code: 403,
        description: "Доступ к ресурсу запрещён."
    },
    NOT_FOUND: {
        code: 404,
        description: "Ресурс не найден."
    },
    METHOD_NOT_ALLOWED: {
        code: 405,
        description: "Метод не поддерживается."
    },
    CONFLICT: {
        code: 409,
        description: "Конфликт запроса с текущим состоянием сервера."
    },
    UNPROCESSABLE_ENTITY: {
        code: 422,
        description: "Ошибка валидации данных."
    },
    INTERNAL_SERVER_ERROR: {
        code: 500,
        description: "Внутренняя ошибка сервера."
    },
    NOT_IMPLEMENTED: {
        code: 501,
        description: "Метод не реализован на сервере."
    },
    BAD_GATEWAY: {
        code: 502,
        description: "Ошибка шлюза."
    },
    SERVICE_UNAVAILABLE: {
        code: 503,
        description: "Сервис временно недоступен."
    },
    GATEWAY_TIMEOUT: {
        code: 504,
        description: "Время ожидания ответа от шлюза истекло."
    }
};

// Accounts
const SQL_CREATE_ACCOUNT = `INSERT INTO accounts (user_id, balance) VALUES (?, 0)`;
const SQL_FIND_ACCOUNT_BY_ID = `SELECT * FROM accounts WHERE id = ?`;
const SQL_FIND_ACCOUNT_BY_USER_ID = `SELECT * FROM accounts WHERE user_id = ?`;
const SQL_UPDATE_ACCOUNT_BALANCE = `UPDATE accounts SET balance = ? WHERE id = ?`;

// Basket
const SQL_CREATE_BASKET = `INSERT INTO basket (user_id) VALUES (?)`;
const SQL_FIND_BASKET_BY_ID = `SELECT * FROM basket WHERE id = ? AND status = 'open'`;
const SQL_FIND_BASKET_BY_USER_ID = `SELECT * FROM basket WHERE user_id = ? AND status = 'open'`;
const SQL_ADD_ITEM_TO_BASKET = `
  INSERT IGNORE INTO basket_items (basket_id, item_id, quantity) 
  VALUES (?, ?, ?) 
  ON DUPLICATE KEY UPDATE quantity = quantity + VALUES(quantity)
`;
const SQL_REMOVE_ITEM_FROM_BASKET = `DELETE FROM basket_items WHERE basket_id = ? AND item_id = ?`;
const SQL_GET_BASKET_ITEMS = `SELECT * FROM basket_items WHERE basket_id = ?`;
const SQL_UPDATE_BASKET_STATUS = `UPDATE basket SET status = ? WHERE id = ?`;

// Warehouse - склад
const SQL_RESERVE_ITEM = `UPDATE warehouse
         SET 
         reserved_quantity = reserved_quantity + ?,
         quantity = quantity - ?
         WHERE product_id = ? AND quantity - reserved_quantity >= ?`;
const RESERVE_ITEM_SUCCESS = 'Резервирование товара выполнено';
const RESERVE_ITEM_FAILED = 'Возникла ошибка при резервирование товара';

const SQL_RELEASE_ITEM = `UPDATE warehouse 
        SET
        reserved_quantity = reserved_quantity - ?,
        quantity = quantity + ?
        WHERE product_id = ?
        AND reserved_quantity - ? >= 0`;
const RELEASE_ITEM_SUCCESS = 'Отмена резервирования товара выполнена';     
const RELEASE_ITEM_FAILED =  'Возникла ошибка при отмене резервирования товара';

const SQL_GET_PRODUCT_BY_PRODUCT_ID = 
        "SELECT product_id, product_name, price, description, quantity, reserved_quantity, created_at, updated_at " +
        "FROM warehouse " +
        "WHERE product_id = ?";
const SQL_GET_PRODUCTS_BY_CATEGORY_ID = 
        "SELECT product_id, product_name, price, description, quantity, reserved_quantity, created_at, updated_at " 
        + "FROM warehouse " 
//        + "WHERE category_id = ? || ? = ?";
const SQL_GET_MEDIAS_BY_PRODUCT_ID = "SELECT media_id, media_key, mime_type, size, is_default from warehouse_media_storage WHERE product_id = ?";

// Delivery
const SQL_RESERVE_COURIER_SLOT = `UPDATE delivery SET reserved = 1 WHERE courier_id = ? AND time_slot = ? AND reserved = 0`;
const SQL_RELEASE_COURIER_SLOT = `UPDATE delivery SET reserved = 0 WHERE courier_id = ? AND time_slot = ?`;

// Orders
const SQL_CREATE_ORDER = `INSERT INTO orders (basket_id, user_id, amount, reference_id, status) VALUES (?, ?, ?, ?, 'pending')`;
const SQL_FIND_ORDER_BY_ID = `SELECT * FROM orders WHERE id = ?`;
const SQL_UPDATE_ORDER_STATUS = `UPDATE orders SET status = ? WHERE id = ?`;


/* Экспорт констант */
module.exports = {
  COMMON_PATH_TO_SITE,
  COMMON_PATH_TO_PAGES,
  COMMON_PATH_TO_PARTIALS,
  COMMON_REGISTRATION_PAGE,
  COMMON_REGISTRATION_CONFIRM_PAGE,
  COMMON_REGISTRATION_SUCCESS_PAGE,
  COMMON_REGISTRATION_DECLINE_PAGE,
  COMMON_REGISTRATION_FAILURE_PAGE,
  COMMON_LOGON_PAGE,
  COMMON_LOGOUT_PAGE,
  COMMON_OUT_SERVICE_PAGE,
  COMMON_SESSION_CLOSE_PAGE,
  COMMON_APP_PAGE,
  COMMON_PRODUCTS_PAGE,
  COMMON_PROFILE_PAGE,
  COMMON_BASKET_PAGE,
  COMMON_ORDERS_PAGE,
  COMMON_GET_ORDER_PAGE,
  COMMON_GET_ORDER_SUCCESS_PAGE,
  COMMON_GET_ORDER_ERROR_PAGE,
  COMMON_404_PAGE , 
  HTTP_CODES,
  SQL_CREATE_ACCOUNT, 

SQL_FIND_ACCOUNT_BY_ID, 
SQL_FIND_ACCOUNT_BY_USER_ID, 
SQL_UPDATE_ACCOUNT_BALANCE, 

SQL_CREATE_BASKET, 
SQL_FIND_BASKET_BY_ID, 
SQL_FIND_BASKET_BY_USER_ID, 
SQL_ADD_ITEM_TO_BASKET, 
SQL_REMOVE_ITEM_FROM_BASKET, 
SQL_GET_BASKET_ITEMS, 
SQL_UPDATE_BASKET_STATUS, 

SQL_RESERVE_ITEM, 
SQL_RELEASE_ITEM, 
SQL_GET_PRODUCT_BY_PRODUCT_ID, 
SQL_GET_PRODUCTS_BY_CATEGORY_ID,

SQL_RESERVE_COURIER_SLOT, 
SQL_RELEASE_COURIER_SLOT, 

SQL_CREATE_ORDER, 
SQL_FIND_ORDER_BY_ID, 
SQL_UPDATE_ORDER_STATUS,
SQL_GET_MEDIAS_BY_PRODUCT_ID,
RESERVE_ITEM_SUCCESS,
RESERVE_ITEM_FAILED,
RELEASE_ITEM_SUCCESS,
RELEASE_ITEM_FAILED
};
