// openfsm-product-media-dto/ProductMediaDTO.js
class ProductMediaDTO {
    constructor({
        id = null,
        productId = null,
        mediaId = null,
        storage = null,
        bucket = null,
        key = null,
        mimeType = null,
        size = null,
        isDefault = 0,
        createdAt = null,
        blockedAt = null,
        deletedAt = null
    } = {}) {
        this.id = id;
        this.productId = productId;
        this.mediaId = mediaId;
        this.storage = storage;
        this.bucket = bucket;
        this.key = key;
        this.mimeType = mimeType;
        this.size = size;
        this.isDefault = isDefault;
        this.createdAt = createdAt;
        this.blockedAt = blockedAt;
        this.deletedAt = deletedAt;
    }

    // Геттеры
    getId() {
        return this.id;
    }

    getProductId() {
        return this.productId;
    }

    getMediaId() {
        return this.mediaId;
    }

    getStorage() {
        return this.storage;
    }

    getBucket() {
        return this.bucket;
    }

    getKey() {
        return this.key;
    }

    getMimeType() {
        return this.mimeType;
    }

    getSize() {
        return this.size;
    }

    getIsDefault() {
        return this.isDefault;
    }

    getCreatedAt() {
        return this.createdAt;
    }

    getBlockedAt() {
        return this.blockedAt;
    }

    getDeletedAt() {
        return this.deletedAt;
    }

    // Сеттеры
    setId(id) {
        this.id = id;
    }

    setProductId(productId) {
        this.productId = productId;
    }

    setMediaId(mediaId) {
        this.mediaId = mediaId;
    }

    setStorage(storage) {
        this.storage = storage;
    }

    setBucket(bucket) {
        this.bucket = bucket;
    }

    setKey(key) {
        this.key = key;
    }

    setMimeType(mimeType) {
        this.mimeType = mimeType;
    }

    setSize(size) {
        this.size = size;
    }

    setIsDefault(isDefault) {
        this.isDefault = isDefault;
    }

    setCreatedAt(createdAt) {
        this.createdAt = createdAt;
    }

    setBlockedAt(blockedAt) {
        this.blockedAt = blockedAt;
    }

    setDeletedAt(deletedAt) {
        this.deletedAt = deletedAt;
    }
}

module.exports = ProductMediaDTO;
