// openfsm-database-connection-producer/DatabaseConnectionProducer.js
const { Pool } = require('pg');
require('dotenv').config();

const pool = new Pool({
  host: process.env.DB_HOST,
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_NAME,
  port: process.env.DB_PORT || 5432, // порт для PostgreSQL, если не указан, используется стандартный
  max: 10, // максимальное количество подключений
  idleTimeoutMillis: 30000, // время ожидания бездействующего соединения
  connectionTimeoutMillis: 2000, // время ожидания соединения
});

const connectWithRetry = (retries = 5, delay = 2000) => {
  return new Promise((resolve, reject) => {
    console.log(process.env.DB_HOST, process.env.DB_USER, process.env.DB_NAME);
    pool.connect((err, client, release) => {
      if (err) {
        console.error('Failed to connect to PostgreSQL database:', err);
        if (retries > 0) {
          console.log(`Retrying to connect in ${delay}ms... (${retries} retries left)`);
          setTimeout(() => {
            connectWithRetry(retries - 1, delay).then(resolve).catch(reject);
          }, delay);
        } else {
          reject('All retries to connect to PostgreSQL database failed');
        }
      } else {
        console.log('Connected to PostgreSQL database SUCCESS');
        release(); // освобождаем соединение
        resolve();
      }
    });
  });
};

// Инициализируем подключение с переподключением
connectWithRetry()
  .then(() => console.log('Database connection established'))
  .catch(error => console.error(error));

module.exports = pool;
