// openfsm-user-helper/UserHelper.js
const db = require('openfsm-database-connection-producer');
const { UserPermissionsDTO, RoleDTO, PermissionDTO } = require('openfsm-permissions-dto');
const { UserDTO } = require('openfsm-user-dto');

exports.create = (email, password, name) => {
  return new Promise((resolve, reject) => {
    const sql = 'INSERT INTO users (email, password, name) VALUES (?, ?, ?)';
    db.query(sql, [email, password, name], (err, result) => {
      (err) 
        ? reject(err)
        : resolve(result.insertId);
    });
  });
};

exports.findByEmail = (email) => {
  return new Promise((resolve, reject) => {
    const sql = 'SELECT * FROM users WHERE email = ?';
    db.query(sql, [email], (err, result) => {
    (err) 
     ? reject(err)
     : resolve(new UserDTO(result[0]));
    });
  });
};

exports.findById = (id) => {
  return new Promise((resolve, reject) => {
    const sql = 'SELECT * FROM users WHERE id = ?';
    db.query(sql, [id], (err, result) => {
      (err)
      ? reject(err)
      : resolve(new UserDTO(result[0]));
    });
  });
};

exports.update = (id, name, email) => {
  return new Promise((resolve, reject) => {
    const sql = 'UPDATE users SET name = ?, email= ? WHERE id = ?';
    db.query(sql, [name, email,  id], (err, result) => {
      (err) 
      ? reject(err)
      : resolve(result);
    });
  });
};


exports.getPermissions = (id) => {
  return new Promise((resolve, reject) => {
    
    let sqlRoles = `
      SELECT DISTINCT(r.name) AS roles, r.description
      FROM users u
      JOIN user_roles ur ON u.id = ur.user_id
      JOIN roles r ON r.id = ur.role_id
      WHERE u.id = ? and u.blocked is null`;
      
    db.query(sqlRoles, [id], (err, roleResult) => {
      if (err) {
        return reject(err);
      }

      // Создание массивов RoleDTO
      let roles = roleResult.map(row => new RoleDTO(row.roles, row.description));
      
      let sqlPermissions = `
        SELECT DISTINCT(p.name) AS permissions, p.description
        FROM users u
        JOIN user_roles ur ON u.id = ur.user_id
        JOIN role_permissions rp ON ur.role_id = rp.role_id
        JOIN permissions p ON rp.permission_id = p.id
        WHERE u.id = ? and u.blocked is null`;

      db.query(sqlPermissions, [id], (err, permissionResult) => {
        if (err) {
          return reject(err);
        }

        // Создание массивов PermissionDTO
        const permissions = permissionResult.map(row => new PermissionDTO(row.permissions, row.description)); 
        
        const userPermissionsDTO = new UserPermissionsDTO(roles, permissions);
        
        resolve({
          roles: userPermissionsDTO.getRoles(), // получение в формате { roleCode: "..." }
          permissions: userPermissionsDTO.getPermissions() // получение в формате { permCode: "..." }
        });
      });
    });
  });
};
