// openfsm-media-image-dto/MediaImageDto.js
class MediaImageDto {
    constructor({ media_id, media_key, mime_type, size, is_default }) {
        this.mediaId = media_id; // camelCase
        this.mediaKey = media_key; // camelCase
        this.mimeType = mime_type; // camelCase
        this.size = size; // camelCase
        this.isDefault = is_default; // camelCase
    }

    // Getter and setter for mediaId
    getMediaId() {
        return this.mediaId;
    }

    setMediaId(mediaId) {
        this.mediaId = mediaId;
    }

    // Getter and setter for mediaKey
    getMediaKey() {
        return this.mediaKey;
    }

    setMediaKey(mediaKey) {
        this.mediaKey = mediaKey;
    }

    // Getter and setter for mimeType
    getMimeType() {
        return this.mimeType;
    }

    setMimeType(mimeType) {
        this.mimeType = mimeType;
    }

    // Getter and setter for size
    getSize() {
        return this.size;
    }

    setSize(size) {
        this.size = size;
    }

    // Getter and setter for isDefault
    getIsDefault() {
        return this.isDefault;
    }

    setIsDefault(isDefault) {
        this.isDefault = isDefault;
    }
}

// Export the class
module.exports = MediaImageDto;
