// openfsm-product-dto/ProductDTO.js
class ProductDto {
    constructor({        
        product_id,
        product_name,
        price,
        description,
        quantity,
        reserved_quantity,
        created_at,
        updated_at,
        blocked_at,
        deleted_at
    } = {}) {        
        this.productId = product_id;
        this.productName = product_name;
        this.price = price;
        this.description = description;
        this.quantity = quantity;
        this.reservedQuantity = reserved_quantity;
        this.createdAt = created_at;
        this.updatedAt = updated_at;
        this.blockedAt = blocked_at;
        this.deletedAt = deleted_at;
    }

    // Getters and setters
    getId() {
        return this.id;
    }

    setId(id) {
        this.id = id;
    }

    getProductId() {
        return this.productId;
    }

    setProductId(productId) {
        this.productId = productId;
    }

    getProductName() {
        return this.productName;
    }

    setProductName(productName) {
        this.productName = productName;
    }

    getPrice() {
        return this.price;
    }

    setPrice(price) {
        this.price = price;
    }

    getDescription() {
        return this.description;
    }

    setDescription(description) {
        this.description = description;
    }

    getQuantity() {
        return this.quantity;
    }

    setQuantity(quantity) {
        this.quantity = quantity;
    }

    getReservedQuantity() {
        return this.reservedQuantity;
    }

    setReservedQuantity(reservedQuantity) {
        this.reservedQuantity = reservedQuantity;
    }

    getCreatedAt() {
        return this.createdAt;
    }

    setCreatedAt(createdAt) {
        this.createdAt = createdAt;
    }

    getUpdatedAt() {
        return this.updatedAt;
    }

    setUpdatedAt(updatedAt) {
        this.updatedAt = updatedAt;
    }

    getBlockedAt() {
        return this.blockedAt;
    }

    setBlockedAt(blockedAt) {
        this.blockedAt = blockedAt;
    }

    getDeletedAt() {
        return this.deletedAt;
    }

    setDeletedAt(blockedAt) {
        this.deletedAt = deletedAt;
    }

}

module.exports = ProductDto;
