// openfsm-basket-helper/BasketHelper.js
// Модель товара в корзине
const db = require('openfsm-database-connection-producer');
const { v4: uuidv4 } = require('uuid'); 
const common       = require('openfsm-common');  /* Библиотека с общими параметрами */
require('dotenv').config();


/* создаем корзину пользователя  */
async function createBasket(userId) {
  const [result] = await db.promise().query(common.SQL_CREATE_BASKET,
//    `INSERT IGNORE INTO warehouse_basket (user_id) VALUES (?)`,
    [userId]
  );
  return result.insertId;
}

/* добавить в корзину товар */
async function addItemToBasket(userId, productId, quantity) {
  let [basket] = await db.promise().query(common.SQL_FIND_BASKET_BY_USER_ID, [userId]); // если нет корзины - создать ее
  if (basket.length === 0) {
    const basketId = await createBasket(userId);
    basket = [{ id: basketId }];
  }

  const basketId = basket[0].id;
  // если есть уже такой зарезервировнный товар - добавляем количество
  const [existingItem] = await db.promise().query( common.SQL_FIND_BASKET_ITEM_BY_PRODUCT_ID,  [basketId, productId]);
  try {
      await db.promise().query( 
        ((existingItem.length > 0)
            ? common.SQL_UPDATE_BASKET_CHANGE_QUANTITY_ITEM_BY_PRODUCT_ID
            : common.SQL_INSERT_BASKET_QUANTITY_ITEM_BY_PRODUCT_ID )
        ,[quantity, basketId, productId]);
    
     return existingItem[0].quantity + quantity;
    } catch (error) {
   return quantity;    
  }
}

/* Удалить запись из корзины */
async function removeItemFromBasket(userId, productId, quantity) {
  let [basket] = await db.promise().query(common.SQL_FIND_BASKET_BY_USER_ID, [userId]); // если нет корзины - создать ее
  if (basket.length === 0) {
    const basketId = await createBasket(userId);
    basket = [{ id: basketId }];  }

    const basketId = basket[0].id;
    const [existingItem] = await db.promise().query( common.SQL_FIND_BASKET_ITEM_BY_PRODUCT_ID,  [basketId, productId]);
    try {
      if (existingItem.length > 0) {
        await db.promise().query( common.SQL_UPDATE_BASKET_CHANGE_QUANTITY_ITEM_BY_PRODUCT_ID, [ -quantity, basketId, productId] );
        return existingItem[0].quantity - quantity;
      } else  return 0;     
    } catch (error) {
   return 0;    
  }
}

/* Получить состояние корзины */
async function getBasket(userId) {
  const [basket] = await db.promise().query(`SELECT id FROM warehouse_basket WHERE user_id = ?`, [userId]);
  if (basket.length === 0) {
    throw new Error('Basket not found');
  }
  const basketId = basket[0].id;
  const [items] = await db.promise().query(`SELECT * FROM warehouse_basket_item WHERE order_id is null and basket_id = ?`, [basketId]);
  return items;
}


/* Сумма товаров в корзине */
async function getBasketAmount(basketId) {
  const [result] =  await db.promise().query(`select sum(bi.quantity*w.price) as amount from warehouse_basket_item bi
        left join warehouse w on (w.product_id = bi.product_id  )
        left join warehouse_basket b on (b.id  = bi.basket_id  )
        where bi.order_id is null and b.id = ?`, [basketId]);
     if (!result[0].amount) {
          throw('Нет товаров в корзине');  
     }
  return result[0].amount;
}

async function getBasketId(userId) {
  const [basket] = await db.promise().query(`SELECT id FROM warehouse_basket WHERE user_id = ?`, [userId]);
  if (basket.length === 0) {
    return null;
  }
  return basket[0].id;
}

/* Закрыть корзины и товарам присвоить orderId*/
async function closeBasket(userId, orderId) {
  try {
    const basketId = await getBasketId(userId)
    if (!basketId) return false;
    await db.promise().query(`UPDATE warehouse_basket_item SET order_id = ? WHERE order_id is null and basket_id = ?`,  [orderId, basketId]);  
   } catch (error) {
    return false;
  }   
  return true; 
}





module.exports = { addItemToBasket, removeItemFromBasket, getBasket, getBasketAmount, closeBasket, getBasketId};
