// common-warehouse-service/common-warehouse-service.js
// Общая библиотека Warehouse
/* SQL Queries */
const SQL = {
  WAREHOUSE: {
    SQL_GET_MEDIAS_BY_PRODUCT_ID : "SELECT media_id, media_key, mime_type, size, is_default from pa_warehouse_service.warehouse_media_storage WHERE product_id = ?",
    SQL_MAIN_MEDIA_BY_PRODUCT_ID : "SELECT media_id, media_key, mime_type, size, is_default from pa_warehouse_service.warehouse_media_storage WHERE product_id = ? LIMIT 0,1",
    SQL_RESERVE_ITEM : `UPDATE  pa_warehouse_service.warehouse SET reserved_quantity = reserved_quantity + ?,  quantity = quantity - ?  WHERE product_id = ? AND quantity - reserved_quantity >= ?`,
    SQL_RELEASE_ITEM : `UPDATE  pa_warehouse_service.warehouse SET reserved_quantity = reserved_quantity - ?,  quantity = quantity + ?  WHERE product_id = ? AND reserved_quantity - ? >= 0`,
    SQL_GET_PRODUCT_BY_PRODUCT_ID : `SELECT product_id, owner_id, product_name, price, description, quantity, reserved_quantity, created_at, updated_at FROM  pa_warehouse_service.warehouse  WHERE product_id = ?`,
    SQL_GET_PRODUCTS_BY_CATEGORY_ID : `SELECT product_id, product_name, price, description, quantity, reserved_quantity, created_at, updated_at FROM  pa_warehouse_service.warehouse `
  },
  BASKET : {  
    SQL_CREATE_BASKET : `INSERT INTO  pa_warehouse_service.warehouse_basket (user_id) VALUES ($1) RETURNING id;`, // создать корзину
    SQL_FIND_BASKET_BY_ID : `SELECT * FROM  pa_warehouse_service.warehouse_basket WHERE id = $1 `,  // получить параметры корзины по id
    SQL_FIND_BASKET_BY_USER_ID : `SELECT * FROM  pa_warehouse_service.warehouse_basket WHERE user_id = $1 `, //  получить параметры корзины по userId
    SQL_FIND_BASKET_ITEM_BY_PRODUCT_ID : `SELECT * FROM  pa_warehouse_service.warehouse_basket_item WHERE  order_id is null and basket_id = $1 AND product_id = $2`,
    SQL_UPDATE_BASKET_CHANGE_QUANTITY_ITEM_BY_PRODUCT_ID : `UPDATE  pa_warehouse_service.warehouse_basket_item SET quantity = quantity + $1 WHERE order_id is null and basket_id = $2 AND product_id = $3 and quantity > 0 RETURNING quantity`,
    SQL_UPDATE_BASKET_REMOVE_ITEM_BY_PRODUCT_ID : `UPDATE  pa_warehouse_service.warehouse_basket_item SET quantity = 0 WHERE order_id is null and basket_id = $1 AND product_id = $2`,
    SQL_INSERT_BASKET_QUANTITY_ITEM_BY_PRODUCT_ID : `INSERT INTO  pa_warehouse_service.warehouse_basket_item (quantity, basket_id, product_id ) VALUES ($1, $2, $3)`,
    SQL_UPSERT_BASKET_QUANTITY : `
    INSERT INTO pa_warehouse_service.warehouse_basket_item (quantity, basket_id, product_id )
    VALUES ($1, $2, $3)
    ON CONFLICT (basket_id, product_id) 
    DO UPDATE SET
         quantity = pa_warehouse_service.warehouse_basket_item.quantity + EXCLUDED.quantity;`,
    SQL_ADD_BASKET_QUANTITY_ITEM_BY_PRODUCT_ID : `INSERT INTO  pa_warehouse_service.warehouse_basket_item (basket_id, product_id, quantity) VALUES ($1, $2, $3)`,
    SQL_GET_BASKET_ITEMS : `SELECT wbi.*, w.product_name, w.price, w.description FROM  pa_warehouse_service.warehouse_basket_item wbi
      LEFT JOIN  pa_warehouse_service.warehouse w on (w.product_id = wbi.product_id) WHERE wbi.quantity > 0 and order_id is null and basket_id = $1`,
    SQL_GET_BASKET_ORDER_ITEMS : `SELECT  wbi.product_id, wbi.quantity, w.product_name, w.price, w.description FROM  pa_warehouse_service.warehouse_basket_item wbi
      LEFT JOIN  pa_warehouse_service.warehouse w on (w.product_id = wbi.product_id)  WHERE wbi.quantity > 0 and order_id = $1 and basket_id = $2`,
    SQL_ORDER_CREATE  : `UPDATE pa_warehouse_service.warehouse_basket_item SET order_id = $1 WHERE order_id is null and basket_id = $2`,
  }
};

const MESSAGES = {
   RU : {    
      INPUT_VALIDATION_ERROR: 'Ошибка проверки входных данных',
      PRODUCT_ID_REQUIRED: 'Ошибка проверки productId',
      PRODUCT_NOT_FOUND: 'Продукт не найден',
      ERROR_FETCHING_MEDIA: 'Ошибка получения медиафалов для продукта',
      ERROR_FETCHING_PRODUCT: 'Ошибка получения продукта',
      ERROR_FETCHING_WAREHOUSE: 'Ошибка при получении информации о состоянии склада',
      RESERVE_ITEM_SUCCESS : 'Резервирование товара выполнено',
      RESERVE_ITEM_FAILED : 'Возникла ошибка при резервирование товара',
      RELEASE_ITEM_SUCCESS : 'Отмена резервирования товара выполнена',   
      RELEASE_ITEM_FAILED :  'Возникла ошибка при отмене резервирования товара',     
      ERROR_REMOVING_PRODUCT_FROM_BASKET : 'Ошибка удаления продукта из корзины',
 
   }
}

/* Экспорт */
module.exports = {
  SQL, MESSAGES
};
