// openfsm-basket-item-dto/BasketItemDto.js
class BasketItemDto {
    /**
     * @param {Object} data - Объект с данными для инициализации.
     * @param {number} data.id - Уникальный идентификатор записи в корзине.
     * @param {number} data.basket_id - Идентификатор корзины.
     * @param {string} data.product_id - Уникальный идентификатор товара.
     * @param {number} data.quantity - Количество товара.
     * @param {number|null} data.order_id - Идентификатор заказа, если товар оформлен.
     * @param {string|null} data.closed_at - Дата и время закрытия (если применимо).
     * @param {string} data.product_name - Название товара.
     * @param {number} data.price - Цена за единицу товара.
     * @param {string} data.description - Описание товара.
     * @param {Array} data.media_files - Массив медиафайлов (ссылки на изображения и др.).
     */
    constructor(data) {        
        this.id = undefined;
        this.basketId = data.basket_id || undefined;
        this.productId = data.product_id || undefined;
        this.quantity = data.quantity || 0;
        this.orderId = data.order_id || undefined;
        this.closedAt = data.closed_at || undefined;
        this.productName = data.product_name || undefined;
        this.price = data.price || 0.0;
        this.description = data.description || undefined;
        this.mediaFiles = data.media_files || [];
    }

    // Геттеры и сеттеры для всех свойств
    getId() {
        return this.id;
    }

    setId(id) {
        this.id = id;
    }

    getBasketId() {
        return this.basketId;
    }

    setBasketId(basketId) {
        this.basketId = basketId;
    }

    getProductId() {
        return this.productId;
    }

    setProductId(productId) {
        this.productId = productId;
    }

    getQuantity() {
        return this.quantity;
    }

    setQuantity(quantity) {
        this.quantity = quantity;
    }

    getOrderId() {
        return this.orderId;
    }

    setOrderId(orderId) {        
         this.orderId = orderId;
    }

    getClosedAt() {
        return this.closedAt;
    }

    setClosedAt(closedAt) {
        this.closedAt = closedAt;
    }

    getProductName() {
        return this.productName;
    }

    setProductName(productName) {
        this.productName = productName;
    }

    getPrice() {
        return this.price;
    }

    setPrice(price) {
        this.price = price;
    }

    getDescription() {
        return this.description;
    }

    setDescription(description) {
        this.description = description;
    }

    getMediaFiles() {
        return this.mediaFiles;
    }

    setMediaFiles(mediaFiles) {
        this.mediaFiles = mediaFiles;
    } 
}

// Экспорт класса
module.exports = BasketItemDto;
